Moving towards a better Polycrystal IC system in phase-field

permcody opened the pull request:
This PR continues to improve the polycrystal IC system in the phase-field module. In particular, I have modified the `PolycrystalUserObjectBase` API to return multiple grains at a point and the FeatureFloodCount has been modified to handle this as well.

The GrainTracker can now "replicate" the state of another `FeatureFloodCount` object which will be useful for unifying the initial grain assignment logic (e.g. The GrainTracker can use the data structures in `PolycrystalUserObjectBase` directly as the initial grain assignments so we'll be able to get rid of all the special-case EBSD logic in the GrainTracker soon and simplify that class).

This PR already works but I don't want to merge it yet. I made need some help figuring out where we should go to make sure we can create "smooth" values on the grain boundaries.

@dschwen - let's chat about this. I'll comment a bit on this PR to show you where to look.

permcody committed:
Update variable indices of stored grains

permcody committed:
Enhance the GrainTracker to extract state from a compatible object

This will allow the GrainTracker to avoid rediscovering the grain structure when
the optional object is supplied. More importantly, this will take care of the
problem of finding grains when elemental information is used to set ICs. This
information can be lost with nodal basis functions.

permcody committed:
First take at creating an EBSD IC on new system

This commit adds the ability to query an EBSD data set using the new
PolycrystalUserObjectBase class for setting up the IC.

permcody committed:
Preparing to flood with overlapping grains

permcody committed:
Modifications to FeatureFloodCount to support multiple IDs at point

This commit makes it so we can locate multiple grains at a single point
in a single pass. This is necessary for building up the IC because we
don't have a small number of disjoint features sitting on a small number
of OPs. We have potentially thousands of grains with a few active at
each point. Also gets the EBSD IC working

refs #8810

permcody commented on modules/phase_field/src/ics/PolycrystalColoringIC.C:
```diff
@@ -36,8 +36,15 @@ PolycrystalColoringIC::PolycrystalColoringIC(const InputParameters & parameters)
 Real
 PolycrystalColoringIC::value(const Point & p)
 {
+  _poly_ic_uo.getGrainsBasedOnPoint(p, _grains);
```
Here's the general Polycrystal IC class that gets what it needs from the UserObject. Here you can see me grabbing a vector of `_grains` at a point.

permcody commented on modules/phase_field/src/ics/PolycrystalColoringIC.C:
```diff
@@ -36,8 +36,15 @@ PolycrystalColoringIC::PolycrystalColoringIC(const InputParameters & parameters)
+    return 0.0;
+
+  // If we have several active grains, we may want to do something different with the OPs here
+  auto grain_id = _grains[0];
```
For now, I'm just arbitrarily grabbing the first grain and using that to figure out which OP to use, then of course the value.

permcody commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -65,6 +66,9 @@ GrainTracker::GrainTracker(const InputParameters & parameters)
     _remap(getParam<bool>("remap_grains")),
     _nl(_fe_problem.getNonlinearSystemBase()),
     _feature_sets_old(declareRestartableData<std::vector<FeatureData>>("unique_grains")),
+    _poly_ic_uo(parameters.isParamValid("polycrystal_ic_uo")
```
Here I optionally couple in the same `PolycrystalUserObject` which I can replicate data from.

permcody commented on modules/phase_field/src/userobjects/EBSDReader.C:
```diff
@@ -262,11 +262,17 @@ EBSDReader::indexFromPoint(const Point & p) const
 
   x_index = (unsigned int)((p(0) - _minx) / _dx);
+  if (x_index == _nx)
+    --x_index;
```
Umm... We'll have to do something about this. The EBSDReader can only return one grain per point.

permcody committed:
Undoing EBSD Reader hack

permcody commented:
@dschwen - This should be ready for review! I think we are pretty close to moving away from the current PolycrystalIC system.

permcody commented on modules/phase_field/src/ics/PolycrystalColoringIC.C:
```diff
@@ -36,9 +36,8 @@ PolycrystalColoringIC::PolycrystalColoringIC(const InputParameters & parameters)
 Real
 PolycrystalColoringIC::value(const Point & p)
```
Very simple IC.

dschwen commented on modules/phase_field/src/userobjects/PolycrystalEBSD.C:
```diff
@@ -0,0 +1,98 @@
+  getGrainsBasedOnPoint(p, grain_ids);
+
+  if (grain_ids.empty())
+    return -1.0;
```
We should never hit this case, but if we do, is returning -1 the best option?

permcody replied:
No you hit it in a multiphase problem. Well... not with this PR but the follow on one where I'm getting the phase stuff all working properly again before trying to use this new object with any of the phase problems. It turns out that you want to build one of these objects per phase instead of trying to read in all the phases in one object. It greatly simplifies the logic in the object. In those cases you might hit a place where there are no active grains on a phase.

dschwen reviewed (dismissed):
I like this design now. Looks good to me (except for the -1 comment)

permcody commented:
This is failing the EBSD + initial adaptivity example. I don't have time to figure that out right now so this is going to rot for a few weeks before I bet back to it. I'll also create a new test once I fix that bug.

permcody committed:
Updating input file but it's currently giving a seg fault.

permcody committed:
EBSD Fixups and new adaptivity test

permcody committed:
Minor changes that should be safe and independent in the FF/GT objects

permcody committed:
EBSD with adaptivity is working but other stuff is broken

moosebuild commented:
Job `Moose Documentation:linux-clang` on f0ccfc2 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9060/site

permcody committed:
Intel is really, really dumb...

moosebuild commented:
Job `Moose Documentation:linux-clang` on a0959f9 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9060/site

permcody commented:
@dschwen - I've fixed this up (and rebased it so it's hard to see what you've looked at and what you haven't. The last three commits, starting at "EBSD Fixups and new adaptivity test" are what's new. This now works with EBSD+AMR and the reconstruction example problem works properly now.

permcody commented:
You can merge this if you'd like. I have another PR that follows on which strips the GrainTracker of all it's EBSD Logic! Also, I'm getting all of the reconstruction tests and 2 phase reconstruction tests working with the new objects. Finally, I'm also dumping `ReconVarIC`.
